### PR TITLE
LLEXT: clean data and code segmentation

### PR DESCRIPTION
**Pull request by lyakh**
Cleanly separate LLEXT code and data segments, handle .bss together with .data. Using Zephyr's automatic section grouping, based on feature flags.

### CODE REVIEW

**Comment by lyakh** [comment]
this also fixes the breakage caused by https://github.com/zephyrproject-rtos/zephyr/pull/73571 and this builds with both Zephyr versions - before and after that change

**Comment by lyakh** [comment]
Before updating this PR to fix sparse warnings, just to record the present CI result: all clean apart from sparse (to be fixed) and jenkins on-device tests. The letter on MTL only had one device unavailable and the other two green, on LNL the result is strange, but must be unrelated https://sof-ci.01.org/sofpr/PR9190/build5052/devicetest/index.html - is this a known issue?

File: src/library_manager/llext_manager.c [hunk 1]
```diff
@@ -108,130 +108,161 @@ static int llext_manager_load_module(uint32_t module_id, const struct sof_man_mo
...
+	/* Read-only data (.rodata, etc.) */
 	void __sparse_cache *va_base_rodata = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
-	size_t st_rodata_size = ctx->segment_size[SOF_MAN_SEGMENT_RODATA];
+		ctx->segment[LIB_MANAGER_RODATA].addr;
+	size_t rodata_size = ctx->segment[LIB_MANAGER_RODATA].size;
+
+	/* Writable data (.data, .bss, etc.) */
+	void __sparse_cache *va_base_data = (void __sparse_cache *)
+		ctx->segment[LIB_MANAGER_RODATA].addr;
```
Comment (kv2019i) [inline src/library_manager/llext_manager.c:203]: @lyakh Shouldn't this be LIB_MANAGER_DATA ? va_base_rodata already set above.

**Review by kv2019i — COMMENTED** [review]
Looks good except one thing in the second-last patch in the series, please see inline.

File: src/library_manager/llext_manager.c (hunk 1 above)
Comment (lyakh) [reply src/library_manager/llext_manager.c:203]: @kv2019i oops, thanks for catching! The clean-up path - easy to miss without wide-scale automated testing...

**Review by kv2019i — APPROVED** [review]
Thanks, good now!

File: src/library_manager/llext_manager.c [hunk 2]
```diff
@@ -51,6 +51,15 @@ extern struct tr_ctx lib_manager_tr;
 
 #define PAGE_SZ		CONFIG_MM_DRV_PAGE_SIZE
 
+static int llext_manager_update_flags(void __sparse_cache *vma, size_t size, uint32_t flags)
+{
+	size_t pre_pad_size = (uintptr_t)vma & (PAGE_SZ - 1);
+	void *aligned_vma = (__sparse_force uint8_t *)vma - pre_pad_size;
```
Comment (softwarecki) [inline src/library_manager/llext_manager.c:57]: ALIGN_DOWN

File: src/library_manager/llext_manager.c [hunk 3]
```diff
@@ -71,177 +80,219 @@ static int llext_manager_align_unmap(void __sparse_cache *vma, size_t size)
...
+		ctx->segment[LIB_MANAGER_BSS].addr;
+	size_t bss_size = ctx->segment[LIB_MANAGER_BSS].size;
 	int ret;
 
+	/* Check, that .bss is within .data */
+	if (bss_size &&
+	    ((uintptr_t)bss_addr + bss_size <= (uintptr_t)va_base_data ||
+	     (uintptr_t)bss_addr >= (uintptr_t)va_base_data + data_size)) {
+		if ((uintptr_t)bss_addr + bss_size == (uintptr_t)va_base_data &&
+		    !((uintptr_t)bss_addr & (PAGE_SZ - 1))) {
```
Comment (softwarecki) [inline src/library_manager/llext_manager.c:143]: ALIGN_DOWN

File: src/library_manager/llext_manager.c [hunk 4]
```diff
@@ -71,177 +80,219 @@ static int llext_manager_align_unmap(void __sparse_cache *vma, size_t size)
...
-
-static int llext_manager_free_module_bss(uint32_t module_id,
-					 const struct sof_man_module *mod)
-{
-	struct lib_manager_mod_ctx *ctx = lib_manager_get_mod_ctx(module_id);
-	size_t bss_size = ctx->segment_size[SOF_MAN_SEGMENT_BSS];
-	void __sparse_cache *va_base = llext_manager_get_bss_address(module_id, mod);
+	ret = llext_manager_align_unmap(va_base_data, data_size);
+	if (ret < 0)
+		return ret;
```
Comment (softwarecki) [inline src/library_manager/llext_manager.c:213]: Shouldn't we try to unmap the `rodata` even if unmapping the `data` fails?

**Review by softwarecki — COMMENTED** [review]
My two cents

File: src/library_manager/llext_manager.c (hunk 2 above)
Comment (lyakh) [reply src/library_manager/llext_manager.c:57]: @softwarecki you could use it, but I also need `pre_pad_size`, so I open code it.

File: src/library_manager/llext_manager.c (hunk 4 above)
Comment (lyakh) [reply src/library_manager/llext_manager.c:213]: @softwarecki ok, this actually might make sense. Will update, thanks!

File: src/library_manager/llext_manager.c (hunk 3 above)
Comment (softwarecki) [reply src/library_manager/llext_manager.c:143]: So you can use `IS_ALIGNED`

Comment (lyakh) [reply src/library_manager/llext_manager.c:143]: I could, but I don't want to use division where masking works too.

**Comment by lyakh** [comment]
CI failures:
- MTL-HDA - not tested
- TGLU-NOCODEC and TGLU-HDA - not tested
- LNLM-HDA - not tested
- LNLM-SDW - https://sof-ci.01.org/sofpr/PR9190/build5178/devicetest/index.html?model=LNLM_SDW_AIOC&testcase=check-playback-3rounds unclear errors - are they known?
- Zephyr main build failures - known regressions
- Zephyr sparse https://github.com/thesofproject/sof/actions/runs/9385669830/job/25844396507?pr=9190 - unrelated error in power.c

**Comment by plbossart** [comment]
> * LNLM-SDW - https://sof-ci.01.org/sofpr/PR9190/build5178/devicetest/index.html?model=LNLM_SDW_AIOC&testcase=check-playback-3rounds unclear errors - are they known?

yes, that's issue https://github.com/thesofproject/linux/issues/4823. Clearly not related to the LLEXT work...

**Review by lgirdwood — APPROVED** [review]
Looks like all comments are addressed. @softwarecki ok for you now ?

**Review by softwarecki — APPROVED** [review]

**Pull request merged**